Perf/avoid final based on query time range

rahul-kumar-saini opened the pull request:
## Overview
- Queries are set as FINAL if they touch projects that currently have unmerged replacements
- This change introduces logic to avoid setting a query as FINAL if the projects it touches have replacements that have a time range before this query's time range
- This is checked using the latest "exclude groups" replacement per project for the query, finding the latest replacement overall and comparing it to the time range of the query

## Changes
- Created new function in `errors_replacer.py` to get the latest timestamp for a replacement given project ids
- Tested said function
- Added a check to see if the query overlaps replacements before setting the query as final

#### Closes SNS-771

fpacifici commented on snuba/datasets/errors_replacer.py:
```diff
@@ -314,6 +314,32 @@ def process_exclude_groups_and_replacement_types_results(
 
+def get_latest_replacement_time_by_projects(
+    project_ids: Sequence[int], state_name: Optional[ReplacerState]
+) -> Optional[float]:
```
please return a datetime. It will be much easier and safe to manage than a bare float. 
It also makes the interface a lot more intuitive. If you see a float in the return value you don't know what it is an you need to read through the code.
If you return a date time you immediately know what it is.

fpacifici commented on snuba/datasets/errors_replacer.py:
```diff
@@ -314,6 +314,32 @@ def process_exclude_groups_and_replacement_types_results(
+    """
+    p = redis_client.pipeline()
+
+    for project_id in project_ids:
+        key, _ = get_project_exclude_groups_key_and_type_key(project_id, state_name)
+        # Gets the item with the highest score for a key, along with the score itself
+        p.zrevrangebyscore(
+            key, float("inf"), float("-inf"), withscores=True, start=0, num=1
+        )
+    # results are in the form [[(str, float)], [(str, float)], ...]
+    results = p.execute()
```
I don't think this is going to be enough to provide the effect you want.
If we add the condition to exclude groups to a query even if that does not overlap with the replacement, the performance issue is minimal with the only exception of queries that are demoted to FINAL because the number of groups to exclude is too high.

What we would like to avoid is to apply FINAL to a query that does not overlap with the replacement.
Which means the redis key you are interested into the is FINAL key (which will need to be provided with a timestamp) not the group exclusion key.

Were you planning to do this as an intermediate step? That would be fine. Otherwise remember there is an additional step to do.

fpacifici replied:
No, you will have to add a timestamp and keep it up to date to the FINAL case as well, to make it work.
If you rely on the TTL, your change won't have any effect as, when the TTL is passed, the key will be removed and FINAL will not be applied anyway.

fpacifici commented on snuba/datasets/storages/processors/replaced_groups.py:
```diff
@@ -43,40 +51,72 @@ def process_query(self, query: Query, request_settings: RequestSettings) -> None
+                query_is_final,
+                exclude_group_ids,
+                replacement_types,
+            ) = get_projects_query_flags(list(project_ids), self.__replacer_state_name,)
```
please remove the comma

fpacifici commented on snuba/datasets/storages/processors/replaced_groups.py:
```diff
@@ -43,40 +51,72 @@ def process_query(self, query: Query, request_settings: RequestSettings) -> None
-                            ),
-                            [Literal(None, p) for p in exclude_group_ids],
+                    assert isinstance(max_group_ids_exclude, int)
+                    if len(exclude_group_ids) > max_group_ids_exclude:
+                        tags["cause"] = "max_groups"
+                        metrics.increment(
+                            "final", tags=tags,
+                        )
+                    assert isinstance(max_group_ids_exclude, int)
+                    if len(exclude_group_ids) > max_group_ids_exclude:
+                        metrics.increment(
+                            "final",
+                            tags={
+                                "cause": "max_groups",
+                                "referrer": request_settings.referrer,
+                                "parent_api": request_settings.get_parent_api(),
+                            },
+                        )
+                        set_final = True
```
This is the second condition twice.

fpacifici commented on snuba/datasets/storages/processors/replaced_groups.py:
```diff
@@ -43,40 +51,72 @@ def process_query(self, query: Query, request_settings: RequestSettings) -> None
-                    metrics.increment(
-                        "final", tags=tags,
+            if not query_is_final and exclude_group_ids:
+                if self._query_overlaps_replacements(query, project_ids):
```
Please invert the condition, the code will be less nested and easier to read sequentially. 
```
if not self._query_overlaps_replacements(...):
    return
```

rahul-kumar-saini replied:
Since that last line of the function still needs to run, it would look like:
```
if not self._query_overlaps_replacements(...):
    query.set_from_clause(replace(query.get_from_clause(), final=False))
    return
...
```
It'll reduce 1 nesting level but it means that `query.set_from_clause()` would need to be duplicated there.

fpacifici reviewed (changes requested):
Please see comments inline. Also update the tests for the ReplacedGroupProcessor as well.

evanh commented on snuba/datasets/errors_replacer.py:
```diff
@@ -314,6 +314,55 @@ def process_exclude_groups_and_replacement_types_results(
+            project_id, state_name
+        )
+        # Gets the item with the highest score for a key, along with the score itself
+        p.zrevrangebyscore(
```
In theory you could do `zrevrange` with `0, 0`. Since our data is sorted by timestamp, the value at the end of the set is by definition the value with the highest timestamp.

However I think it would be better to maybe try to move this logic behind the existing redis call.

evanh commented on snuba/datasets/storages/processors/replaced_groups.py:
```diff
@@ -71,12 +83,27 @@ def process_query(self, query: Query, request_settings: RequestSettings) -> None
+        or not this Query's time range overlaps with any replacement for any of the
+        projects.
+        """
+        query_from, _ = get_time_range(query, "timestamp")
```
I don't think this logic is correct. Shouldn't we be comparing `query_to` instead of `query_from`? Then the check would be `return latest_replacement_time <= query_to` or "did the replacement happen before or after the time period the query is querying".

rahul-kumar-saini replied:
From my understanding, a replacement could be touching data an unknown amount of time back. So if a replacement happens at a certain time, any query looking at data before that time would be "overlapping". For now, a query doesn't need final if it's looking at data that has no replacements yet.

`.......Time ->....`


Query doesn't overlap:

`...Rep..From..To..` <- This case would say the query does overlap if we `return latest <= query_to`


Query overlaps:

`..From..Rep..To...` 
`.From..To..Rep....` <- This case would say the query doesn't overlap if we `return latest <= query_to`

If my understanding is incorrect, I can change the logic.

rahul-kumar-saini commented on snuba/datasets/errors_replacer.py:
```diff
@@ -259,6 +266,14 @@ def get_projects_query_flags(
         p.zrevrangebyscore(type_key, float("inf"), now - settings.REPLACER_KEY_TTL)
 
+    for exclude_groups_key, _ in exclude_groups_keys_and_types:
+        p.zrevrange(
```
This `zrevrange` gets the latest item only and has `withscores=True` so we can get its score. The other call is getting all items but without the scores. I could change it so that call also responds with the scores but it would be a lot of extra data we don't need since only the latest score matters.

rahul-kumar-saini commented:
In order to keep track of when a "project needs final" replacement occurs, I swapped out `<key>:True` for `<key>:<timestamp>` in `set_project_needs_final()`. I've added a backwards compatibility check in the code which can be removed later.

I've also moved the overlap check to the beginning of `process_query()`.

vladkluev commented on snuba/datasets/errors_replacer.py:
```diff
@@ -206,93 +213,129 @@ def set_project_needs_final(
     - `groups_replacement_types` slice is `len_projects * 2` long
+    - `latest_exclude_groups_replacements` slice is `len_projects` long
+
+    [needs_final   |exclude_groups    |...]
+    [ len(projects)|len(projects) * 2 |   ]
```
either remove this explanation or replace the above one with what it actually is

evanh commented on snuba/datasets/errors_replacer.py:
```diff
@@ -206,93 +213,129 @@ def set_project_needs_final(
     ]
-
     for exclude_groups_key, _ in exclude_groups_keys_and_types:
+        # remove stale excluded groups
```
Nit: some of these comments of seem unnecessary.

fpacifici replied:
to echo what Evan said, commenting to explain what the code does generally means your code is not readable enough. Fixing the code is the preferred option instead of adding comments. So if you feel you needed to explain that, probably refactoring is needed. Here I doubt that comment is needed entirely.

Commenting on the "why" the code does something is different and can be useful a lot. Basically the idea is that the comment provides context, it does not explains what we can already read in the code.

In this specific case, I think you are explaining everything in the docstring. No need to pollute the code, if you want to add more context, please feel free to beef up the docstring

vladkluev replied:
I think this comment is better here than in the docstring as it actually tells me what it's doing when it's doing it. I also think that comment is actually communicating intent. Consider this line:

```python3
p.zremrangebyscore(
    exclude_groups_key, float("-inf"), now - settings.REPLACER_KEY_TTL
)
``` 

In order to understand what it's doing you need to:
* Know what `remrangebyscore` does
* its arguments and their order
* how this sorted set is scored

I understand that somebody with a lot of context in the codebase can see this pattern clearly but processing the above without the comment is much harder than with.

fpacifici commented on snuba/datasets/errors_replacer.py:
```diff
@@ -206,93 +213,124 @@ def set_project_needs_final(
+    5. Fetch latest exclude groups replacement type for each Project
 
-    Returns (needs_final, group_ids_to_exclude, replacement_types)
+    Returns (
+        needs_final,
+        group_ids_to_exclude,
+        replacement_types,
+        latest_replacement_time
+    )
```
This does not add any information with respect to the return type structure. You do not need to explain that here, and it can be counter productive if we update the type and forget to update the comment.

Remove please.

fpacifici commented on snuba/datasets/errors_replacer.py:
```diff
@@ -311,7 +349,45 @@ def process_exclude_groups_and_replacement_types_results(
+        latest_need_final_replacement_times = [
+            # Backwards compatibility: Before it was simply "True" at each key,
+            # now it's the timestamp at which the key was added.
+            # This should stay for at least 24h after this change is deployed.
+            # `and timestamp != b"True"` can be removed when there are no
+            # longer any "key:True" entries
```
This has to stay indefinitely as you do not know when single tenant and self hosted deployment will upgrade and use this version. It may happen a year from now.

fpacifici commented on snuba/datasets/errors_replacer.py:
```diff
@@ -168,25 +168,32 @@ def set_project_exclude_groups(
     replacement_type: ReplacementType,
 ) -> None:
     """
+    This method is called when a replacement comes in. For a specific project, record
+    the group ids which were deleted as a result of this replacement
+
```
ok, this time, though, going on, I would discourage you from mixing cosmetics that are unrelated with your PR with the logic of your PR.
That makes reviewing harder because it means spending time to understand why a given change is there.

fpacifici replied:
Then in such case there should be first a PR with the refactoring and then a PR with the logic change.

fpacifici commented on snuba/datasets/errors_replacer.py:
```diff
@@ -206,93 +213,124 @@ def set_project_needs_final(
 
 
-def process_exclude_groups_and_replacement_types_results(
+def _process_exclude_groups_and_replacement_types_results(
```
This method and the following seem to have the only goal of producing the the `ProjectsQueryFlags` instance from the bare results you got from Redis. 

Now the `ProjectQueryFlags` data class is just a bag of data. It seems these methods should be part of that class as a static factory method. 

```
class ProjectsQueryFlags:
....
@classmethod
def load_from_redis(cls, project_ids: Sequence[int], state_name: Optional[ReplacerState]):  ProjectsQueryFlags:
   ...the whole logic ...
```

This encapsulates these methods into a class that is the only place concerned with them.

Please do that in a separate PR

rahul-kumar-saini replied:
#2155

fpacifici commented on snuba/datasets/storages/processors/replaced_groups.py:
```diff
@@ -43,40 +51,64 @@ def process_query(self, query: Query, request_settings: RequestSettings) -> None
         if project_ids:
-            final, exclude_group_ids, replacement_types = get_projects_query_flags(
-                list(project_ids), self.__replacer_state_name,
+            flags: ProjectsQueryFlags = get_projects_query_flags(
```
the PR does not test the change to this query processor. Please let's cover the new case
https://github.com/getsentry/snuba/blob/master/tests/datasets/storages/processors/test_replaced_groups.py

fpacifici commented on snuba/datasets/storages/processors/replaced_groups.py:
```diff
@@ -43,40 +51,64 @@ def process_query(self, query: Query, request_settings: RequestSettings) -> None
-                assert isinstance(max_group_ids_exclude, int)
-                if len(exclude_group_ids) > max_group_ids_exclude:
-                    tags["cause"] = "max_groups"
+            # Continue process if the query's time range overlaps with any replacements
```
The code is clear here. Please remove the comment. If you want you can add some context in the docstring.

fpacifici commented on snuba/datasets/storages/processors/replaced_groups.py:
```diff
@@ -43,40 +51,64 @@ def process_query(self, query: Query, request_settings: RequestSettings) -> None
-                if len(exclude_group_ids) > max_group_ids_exclude:
-                    tags["cause"] = "max_groups"
+            # Continue process if the query's time range overlaps with any replacements
+            if self._query_overlaps_replacements(query, flags.latest_replacement_time):
```
Please invert this condition and exit early if the query does not overlap the replacements. This reduces the indentation of the code making it more readable.

```
if not self._query_overlaps_replacements(query, flags.latest_replacement_time):
      query.set_from_clause(replace(query.get_from_clause(), final=False))
      return

# Now add your logic
```

This also makes the code easier to review as the entire if block is not part of the changeset.

fpacifici commented on snuba/datasets/errors_replacer.py:
```diff
@@ -206,93 +213,124 @@ def set_project_needs_final(
             exclude_groups_key, float("inf"), now - settings.REPLACER_KEY_TTL
         )
 
+    # get the replacement types for metrics purposes
```
please remove

vladkluev replied:
why? it's communicating intent

fpacifici reviewed (changes requested):
Please add tests for the query processor changes. 

Otherwise just minor comments

fpacifici commented on tests/datasets/storages/processors/test_replaced_groups.py:
```diff
@@ -32,17 +33,68 @@ def build_in(project_column: str, projects: Sequence[int]) -> Expression:
     redis_client.flushdb()
 
 
+def test_query_set_final(query: ClickhouseQuery) -> None:
+    enforcer = PostReplacementConsistencyEnforcer("project_id", None)
+
+    enforcer._set_query_final(query, True)
+    assert query.get_from_clause().final
+
+    enforcer._set_query_final(query, False)
+    assert not query.get_from_clause().final
```
No need to test what is not a public method. 
Private/protected methods are considered internal implementation details. 
You generally rely on unit test to ensure no regression if you do a refactoring that keeps the interface unchanged. 
You run the test as it is and it must pass after the refactoring as well as before.

If your unit test tests individual private/protected methods, that cannot be done, because you will have to make changes to the test if you refactor the internals of the class. 
And when you need to make changes to the test to test for regressions you may not have introduced regressions in the test itself.

fpacifici commented on tests/datasets/storages/processors/test_replaced_groups.py:
```diff
@@ -127,3 +173,53 @@ def test_too_many_groups_to_exclude(query: ClickhouseQuery) -> None:
+        ReplacerState.EVENTS,
+        ReplacementType.EXCLUDE_GROUPS,  # Arbitrary replacement type, no impact on tests
+    )
+    enforcer._set_query_final(query_with_timestamp, True)
```
Please do not set the flag to True before the test. This way you are not testing the behavior of the `process_query` method anymore as you don't know if the final flag is on because it was set by the process_query method or by yourself on this line.

fpacifici commented on tests/datasets/storages/processors/test_replaced_groups.py:
```diff
@@ -127,3 +173,53 @@ def test_too_many_groups_to_exclude(query: ClickhouseQuery) -> None:
     assert query.get_from_clause().final
+
+
+def test_query_overlaps_replacements_method(
+    query: ClickhouseQuery, query_with_timestamp: ClickhouseQuery,
+) -> None:
+    enforcer = PostReplacementConsistencyEnforcer("project_id", None)
+    # query from < latest replacement time < query to
+    assert enforcer._query_overlaps_replacements(
+        query_with_timestamp, datetime(2021, 1, 1, 10)
+    )
+    # query to < latest replacement time
+    assert enforcer._query_overlaps_replacements(
+        query_with_timestamp, datetime(2021, 2, 1)
+    )
+    # latest replacement time < query from
+    assert not enforcer._query_overlaps_replacements(
+        query_with_timestamp, datetime(2020, 1, 1)
+    )
+    # replacement time unknown, default to overlaps
+    assert enforcer._query_overlaps_replacements(query_with_timestamp, None)
+    # query time range unknown, default to overlaps
+    assert enforcer._query_overlaps_replacements(query, datetime(2021, 2, 1))
```
You can skip this test as well for the same reason above. 
If you think the method is interesting enough to be tested on its own, take it out of the class as a stand alone function and let it be exposed. For as long as it is a protected method, it should be considered implementation detail of the class and its hierarchy.

fpacifici reviewed (approved):
Please address the comments on the test before merging together with those on the comments.
